src/callbacks/wandb_callbacks.py

```diff
@@ -41,18 +51,38 @@ def on_train_start(self, trainer, pl_module):
 
 
 class UploadCodeAsArtifact(Callback):
-    """Upload all *.py files to wandb as an artifact, at the beginning of the run."""
+    """Upload all code files to wandb as an artifact, at the beginning of the run."""
+
+    def __init__(self, code_dir: str, use_git: bool = True):
+        """
 
-    def __init__(self, code_dir: str):
+        :param code_dir:
+        :param use_git: if using git, then upload all files that are not ignored by git. if not using git, then upload all '*.py' file
+        """
         self.code_dir = code_dir
+        self.use_git = use_git
 
+    @rank_zero_only
     def on_train_start(self, trainer, pl_module):
         logger = get_wandb_logger(trainer=trainer)
         experiment = logger.experiment
 
         code = wandb.Artifact("project-source", type="code")
-        for path in glob.glob(os.path.join(self.code_dir, "**/*.py"), recursive=True):
-            code.add_file(path)
+
+        if self.use_git:
+            # get .git folder
+            # https://alexwlchan.net/2020/11/a-python-function-to-ignore-a-path-with-git-info-exclude/
+            git_dir_path = Path(subprocess.check_output(["git", "rev-parse", "--git-dir"]).strip().decode("utf8")).resolve()
+
+            for path in Path(self.code_dir).rglob('*'):
+                if (path.is_file()
+                        and (not is_relative_to(path, git_dir_path))  # ignore files in .git
+                        and (subprocess.run(['git', 'check-ignore', '-q', str(path)]).returncode == 1)):  # ignore files ignored by git
+                    code.add_file(str(path), name=str(path.relative_to(self.code_dir)))
+
+        else:
+            for path in Path(self.code_dir).rglob('*.py'):
+                code.add_file(str(path), name=str(path.relative_to(self.code_dir)))
 
         experiment.use_artifact(code)
```

Comment:
For some reason the run view doesn't show me its code artifact when using `use_artifact` but it works correctly with `log_artifact`. Although it always shows up correctly in global view.
Do you also have this issue?
Maybe this callback also should be using `log_artifact` for now

Reply:
In my experience, sometimes the artifacts do not show up during the run but will eventually show up after the run is finished